include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2020 Matt Borland
+//
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+#define BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+
+#include <boost/math/special_functions/prime.hpp>
+#include <boost/assert.hpp>
+#include <deque>
+#include <vector>
+#include <iterator>
+#include <iostream>
+
+namespace boost { namespace math
+{
+
+// https://mathworld.wolfram.com/SieveofEratosthenes.html
+// https://www.cs.utexas.edu/users/misra/scannedPdf.dir/linearSieve.pdf
+template<class Z, class OutputIterator>
+auto prime_sieve(Z lower_bound, Z upper_bound, OutputIterator output) -> decltype(output)
+{
+    static_assert(std::is_integral<Z>::value, "No primes for floating point types");
+    BOOST_ASSERT_MSG(upper_bound + 1 < std::numeric_limits<Z>::max(), "Type Overflow");
+    std::vector<Z> least_divisors(upper_bound + 1, 0);
+    std::deque<Z> primes;
```

Comment:
You might want to use the prime number theorem to reserve the correct amount of memory: π(x) = x/ln(x).

Reply:
@NAThompson There is no `std::deque::reserve()` as it does not exist in contiguous memory like a `std::vector`. `std::deque` is faster here for primes than a `std::vector` because insertion time into a `std::deque` is O(1) as opposed to O(N) for a `std::vector`.

Reply:
But you know almost exactly how big the vector will be because of the prime number theorem.

Reply:
@NAThompson Is there an implementation of ln somewhere in the code base that supports `boost::multiprecision::cpp_int`? A lot of errors are thrown by `boost::math::tools::promote_args` using `boost::math::log1p()`. Also do you know multiprecision is constructed at fixed size, or is it variable?

Reply:
You will need to explicitly cast a cpp_int to type double to pass through `log`/`log1p`.

For built-in integer types these are promoted to double automatically (originally for compatibility with C99), but calling a floating point function with a multiprecision integer is always an error: IMO rightly so as this is normally a mistake ;)